File: web/src/lib/taskLaunch.ts
/**
 * One owner for the AutomationTask payload contract — the TypeScript mirror of
 * `src/boss_agent/task_launch.py`.
 *
 * The payload had no owner: the launch modal built it three ways inline, the searches
 * page had a fourth copy, the jobs page a fifth, and `rerunTask` spread-copied
 * whatever the original carried. Their defaults had already diverged — `min_score`
 * was 75 in the modal and 70 in the searches trigger and absent from "run scheduled
 * now", and preview depth differed per caller.
 *
 * Both languages are pinned by `config/task_launch.cases.json`, so the builder cannot
 * drift from the worker's expectation.
 */

/** The baseline relevance threshold, declared once. */
export const MIN_SCORE = 70;

/** Baseline job ceiling for a search dispatch. */
export const DEFAULT_MAX_JOBS = 30;

/** Seconds the worker previews a drafted greeting before moving on. */
export const DEFAULT_PREVIEW_TIMEOUT_SEC = 3.0;

/** The launch payload key carrying task provenance. */
export const SOURCE_KEY = 'source';

/** Task Provenance — where a task came from (CONTEXT.md). */
export type LaunchSource = 'manual' | 'test' | 'scheduler';

/**
 * How deep a launch may go.
 *
 * `undefined` is the third, distinct state: *honour the configured* `chat.dry_run`.
 * A caller that omits the mode lets the operator's drill setting win, which is what
 * the searches-page trigger has always meant.
 */
export type LaunchMode = 'draft' | 'live' | undefined;

export type LaunchKind = 'search' | 'chat_cleanup' | 'login_diagnostic';

export type TaskTypeName = 'SCRAPE_JOBS' | 'AUTO_APPLY' | 'CHECK_CHAT' | 'CHECK_LOGIN';

export type TargetActionName = 'save_jd' | 'auto_apply';

export interface TaskLaunch {
	task_type: TaskTypeName;
	payload: Record<string, unknown>;
}

/** The SavedSearch fields the launch contract reads. */
export interface SearchLaunchInput {
	id: string;
	name?: string;
	keyword?: string;
	enable_search?: boolean;
	enable_filter?: boolean;
	// Opaque to the contract: the builder forwards it, and the two languages'
	// normalizations of a nested filter are deliberately not compared.
	filter?: unknown;
	target_action?: TargetActionName | string;
	target_task_type?: string;
	max_jobs?: number;
	screening_policy?: unknown;
}

/** The resolved `chat:` block the cleanup launch carries. */
export interface ChatAcknowledgment {
	rejection_reply_text?: string;
	max_scan_depth?: number;
	dry_run?: boolean;
}

export class LaunchContractError extends Error {}

function targetActionFor(search: SearchLaunchInput): TargetActionName {
	const declared = search.target_action;
	if (declared === 'auto_apply' || declared === 'save_jd') {
		return declared;
	}
	if (declared) {
		throw new LaunchContractError(
			`SavedSearch ${search.id} declares an unknown target_action ${declared}; expected save_jd or auto_apply`
		);
	}
	return search.target_task_type === 'AUTO_APPLY' ? 'auto_apply' : 'save_jd';
}

/**
 * The (preview_only, auto_send) pair the handlers consume today. Wire keys stay
 * stable for rollout; only who computes them changes. A save-only search is preview
 * by definition — it has nothing to send.
 */
function previewFlags(action: TargetActionName, mode: LaunchMode): [boolean, boolean] {
	if (action !== 'auto_apply') return [true, false];
	if (mode === 'live') return [false, true];
	return [true, false];
}

export function buildSearchLaunch(
	search: SearchLaunchInput,
	options: {
		source: LaunchSource;
		mode?: LaunchMode;
		minScore?: number;
		// Opaque too: forwarded to the worker as the payload's candidate block.
		candidateProfile?: unknown;
	}
): TaskLaunch {
	const action = targetActionFor(search);
	const [previewOnly, autoSend] = previewFlags(action, options.mode);

	const payload: Record<string, unknown> = {
		saved_search_id: search.id,
		search_id: search.id,
		search_name: search.name ?? '',
		keyword: search.keyword || '',
		enable_search: search.enable_search !== false,
		enable_filter: search.enable_filter !== false,
		filter: search.filter || {},
		target_action: action,
		max_jobs: search.max_jobs || DEFAULT_MAX_JOBS,
		min_score: options.minScore === undefined ? MIN_SCORE : Number(options.minScore),
		preview_only: previewOnly,
		auto_send: autoSend,
		preview_timeout_sec: DEFAULT_PREVIEW_TIMEOUT_SEC,
		[SOURCE_KEY]: options.source
	};
	if (search.screening_policy) payload.screening_policy = search.screening_policy;
	if (options.candidateProfile) payload.candidate_profile = options.candidateProfile;
	return { task_type: action === 'auto_apply' ? 'AUTO_APPLY' : 'SCRAPE_JOBS', payload };
}

export function buildChatCleanupLaunch(options: {
	source: LaunchSource;
	search?: SearchLaunchInput | null;
	mode?: LaunchMode;
	chat?: ChatAcknowledgment | null;
}): TaskLaunch {
	const chat = options.chat || {};
	const payload: Record<string, unknown> = {
		// One convention, not three. `mode` decides when the caller states one; when it
		// does not, the *configured* drill mode wins — the same resolution the worker
		// applies, so a launch can no longer be read three different ways depending on
		// which surface produced it.
		dry_run: options.mode === undefined ? chat.dry_run === true : options.mode === 'draft',
		rejection_reply_text: chat.rejection_reply_text ?? '',
		max_scan_depth: chat.max_scan_depth ?? 0,
		[SOURCE_KEY]: options.source
	};
	if (options.search) {
		payload.saved_search_id = options.search.id;
		payload.search_id = options.search.id;
		payload.search_name = options.search.name ?? '';
	}
	return { task_type: 'CHECK_CHAT', payload };
}

/**
 * The CHECK_LOGIN probe. It carries provenance and nothing else: the handler reads no
 * payload, and the `mode: 'diagnostic'` key it used to be given had no reader.
 */
export function buildLoginDiagnosticLaunch(options: { source: LaunchSource }): TaskLaunch {
	return { task_type: 'CHECK_LOGIN', payload: { [SOURCE_KEY]: options.source } };
}

export function buildLaunch(
	kind: LaunchKind,
	options: {
		source: LaunchSource;
		search?: SearchLaunchInput | null;
		mode?: LaunchMode;
		chat?: ChatAcknowledgment | null;
		minScore?: number;
		candidateProfile?: Record<string, unknown> | null;
	}
): TaskLaunch {
	if (kind === 'search') {
		if (!options.search) throw new LaunchContractError('a search launch requires a SavedSearch');
		return buildSearchLaunch(options.search, options);
	}
	if (kind === 'chat_cleanup') return buildChatCleanupLaunch(options);
	if (kind === 'login_diagnostic') return buildLoginDiagnosticLaunch(options);
	throw new LaunchContractError(`unknown task kind ${kind}`);
}

/**
 * Inputs a rerun must carry over verbatim.
 *
 * They are launch *inputs*, not contract output — the builder has no way to re-derive
 * them from a previous payload — so they are named explicitly rather than
 * spread-copied. A blind spread is what let a stale `min_score` or an inverted
 * preview flag survive into a rerun.
 */
const RERUN_CARRIED_INPUTS = [
	'direct_job_id',
	'job_title',
	'company_name',
	'candidate_profile',
	'screening_policy'
] as const;

/** The prior launch, as a rerun needs to read it. */
export interface RerunSource {
	task_type: string;
	payload: Record<string, any>;
}

/**
 * Rebuild a task's payload from the original's *inputs*, through the builder.
 *
 * The contract fields are re-derived, so a divergence carried by the original — an
 * old default, a flag from a builder that no longer exists — cannot propagate. Only
 * the fields the builder cannot derive are carried over, by name.
 */
export function rebuildRerunPayload(
	original: RerunSource,
	rerunOf: string,
	source: LaunchSource = 'manual'
): Record<string, unknown> {
	const prior = original.payload || {};

	if (original.task_type === 'CHECK_CHAT') {
		const search =
			prior.saved_search_id || prior.search_id
				? { id: String(prior.saved_search_id ?? prior.search_id), name: prior.search_name }
				: null;
		const launch = buildChatCleanupLaunch({
			source,
			search,
			// No mode: a rerun honours the configured drill mode, like the scheduler does.
			mode: undefined,
			chat: {
				rejection_reply_text: prior.rejection_reply_text,
				max_scan_depth: prior.max_scan_depth,
				dry_run: prior.dry_run
			}
		});
		return { ...launch.payload, rerun_of: rerunOf };
	}

	if (original.task_type === 'CHECK_LOGIN') {
		return { ...buildLoginDiagnosticLaunch({ source }).payload, rerun_of: rerunOf };
	}

	const live = prior.preview_only === false && prior.auto_send === true;
	const launch = buildSearchLaunch(
		{
			id: String(prior.saved_search_id ?? prior.search_id ?? ''),
			name: prior.search_name,
			keyword: prior.keyword,
			enable_search: prior.enable_search,
			enable_filter: prior.enable_filter,
			filter: prior.filter,
			target_action: prior.target_action,
			max_jobs: prior.max_jobs
		},
		{ source, mode: live ? 'live' : 'draft' }
	);

	const carried: Record<string, unknown> = {};
	for (const key of RERUN_CARRIED_INPUTS) {
		if (prior[key] !== undefined) carried[key] = prior[key];
	}
	return { ...launch.payload, ...carried, rerun_of: rerunOf };
}

File: src/boss_agent/task_launch.py
"""
boss_agent.task_launch
======================
One owner for the ``AutomationTask`` payload contract.

The payload is the wire contract between the Web Dashboard / Automation Scheduler /
Startup Rejection Cleanup Barrier on one side and the Task Handler Strategy on the
other. It had no owner: at least seven builders produced it with already-diverged
semantics, and nothing validated it.

The divergences were not cosmetic. ``min_score`` defaulted to 75 in the launch modal
and 70 in the scheduler, and was simply absent from the dashboard's "run scheduled
now" button. ``preview_only`` was ``True`` from every web builder and ``False`` from
the scheduler, so the *same* SavedSearch produced inverted execution depth depending
on who dispatched it. And task provenance — a first-class CONTEXT.md attribute with
defined ``manual``/``test``/``scheduler`` semantics — existed nowhere in code, faked
instead with payload markers like ``startup_cleanup`` and ``scheduled``.

This module turns ``(kind, provenance, search, mode)`` into a validated payload. The
defaults are declared once, the wire keys stay stable for rollout, and
``config/task_launch.cases.json`` pins the output so the TypeScript builder cannot
drift from it.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any

from .broker.collection_schema import SAVED_SEARCH_MAX_JOBS
from .broker.models import TaskType
from .models import SavedSearch, TargetAction, TargetTaskType
from .settings import resolve_chat_acknowledgment_settings

#: The baseline relevance threshold. One value, so a manual launch and a scheduled run
#: of the same SavedSearch cannot disagree about which jobs qualify.
MIN_SCORE = 70

#: Baseline job ceiling for a search dispatch. An alias, not a restatement: the
#: Collection Schema declares the `saved_searches.max_jobs` default, and a second
#: literal here is how 20-vs-30 drifted apart in the first place.
DEFAULT_MAX_JOBS = SAVED_SEARCH_MAX_JOBS

#: Seconds the worker previews a drafted greeting before moving on.
DEFAULT_PREVIEW_TIMEOUT_SEC = 3.0


class LaunchSource(StrEnum):
    """Task Provenance — where a task came from, as CONTEXT.md defines it.

    A real attribute on the task record rather than a payload marker: startup
    reclamation cancels ``TEST`` tasks so an automated suite never contends with the
    live worker for the device, and the dashboard shows the rest.
    """

    MANUAL = "manual"
    TEST = "test"
    SCHEDULER = "scheduler"


class LaunchMode(StrEnum):
    """How deep a launch is allowed to go.

    Two modes, not three: "dry run" and "draft" were the same wire intent expressed in
    two vocabularies — compute it, show it, do not send it — and collapsing them is the
    whole point of having one builder. ``None`` is a third, distinct state: *honour the
    configured* ``chat.dry_run``, which is what the searches-page trigger has always
    meant by omitting the key.
    """

    DRAFT = "draft"
    LIVE = "live"


class TaskKind(StrEnum):
    """The Task Handler Strategy a launch targets."""

    SEARCH = "search"
    CHAT_CLEANUP = "chat_cleanup"
    LOGIN_DIAGNOSTIC = "login_diagnostic"


class LaunchContractError(ValueError):
    """A launch request the worker could not honour as written."""


def _target_action_for(search: SavedSearch) -> TargetAction:
    """The SavedSearch's configured execution depth, resolved the same way everywhere.

    A malformed `target_action` is rejected here rather than silently defaulting: a
    typo used to fall through to `SAVE_JD`, quietly turning a greeting run into a
    save-only pass.
    """
    raw = search.target_action or (
        TargetAction.AUTO_APPLY
        if search.target_task_type == TargetTaskType.AUTO_APPLY
        else TargetAction.SAVE_JD
    )
    try:
        action = TargetAction(raw)
    except ValueError as exc:
        raise LaunchContractError(
            f"SavedSearch {search.id!r} declares an unknown target_action {raw!r}; "
            f"expected one of {[a.value for a in TargetAction]}"
        ) from exc
    if action not in (TargetAction.AUTO_APPLY, TargetAction.SAVE_JD):
        raise LaunchContractError(
            f"SavedSearch {search.id!r} declares an unsupported target_action {raw!r}; "
            f"expected one of {[a.value for a in TargetAction]}"
        )
    return action


@dataclass(frozen=True)
class TaskLaunch:
    """A validated launch: the task type, the payload, and where it came from.

    Provenance travels as a *task attribute*, not a payload key: the worker's startup
    sweep and the dashboard both need to read it, and neither should have to do payload
    archaeology for a field the record already has a column for.
    """

    task_type: TaskType
    payload: dict[str, Any] = field(default_factory=dict)
    source: LaunchSource = LaunchSource.MANUAL


def build_search_launch(
    search: SavedSearch,
    *,
    source: LaunchSource,
    mode: LaunchMode | None = LaunchMode.DRAFT,
    candidate_profile: dict[str, Any] | None = None,
    min_score: int | None = None,
) -> TaskLaunch:
    """Launch a saved search as a scrape or an auto-apply run.

    ``min_score`` is a caller *choice*, not a competing default: the parameter exists so
    the modal can offer a stricter threshold than the baseline, and the baseline is what
    everyone else gets. Leaving it unset used to mean three different numbers.
    """
    action = _target_action_for(search)
    task_type = (
        TaskType.AUTO_APPLY if action == TargetAction.AUTO_APPLY else TaskType.SCRAPE_JOBS
    )

    # A save-only search never greets, so its preview flags are not a caller choice.
    preview_only, auto_send = _preview_flags(action, mode)

    search_dict = search.to_dict()
    payload: dict[str, Any] = {
        "saved_search_id": search.id,
        "search_id": search.id,
        "search_name": search.name,
        "keyword": search_dict.get("keyword") or "",
        "enable_search": search.enable_search,
        "enable_filter": search.enable_filter,
        "filter": search_dict.get("filter") or {},
        "target_action": action.value,
        "max_jobs": search.max_jobs or DEFAULT_MAX_JOBS,
        "min_score": MIN_SCORE if min_score is None else int(min_score),
        "preview_only": preview_only,
        "auto_send": auto_send,
        "preview_timeout_sec": DEFAULT_PREVIEW_TIMEOUT_SEC,
    }
    if search_dict.get("screening_policy"):
        payload["screening_policy"] = search_dict["screening_policy"]
    if candidate_profile:
        payload["candidate_profile"] = candidate_profile
    return TaskLaunch(task_type=task_type, payload=payload, source=source)


def _preview_flags(action: TargetAction, mode: LaunchMode | None) -> tuple[bool, bool]:
    """The (preview_only, auto_send) pair the handlers consume today.

    Wire keys stay stable for rollout; only who computes them changes. A save-only
    search is preview by definition — it has nothing to send.
    """
    if action != TargetAction.AUTO_APPLY:
        return True, False
    if mode == LaunchMode.LIVE:
        return False, True
    return True, False


def build_chat_cleanup_launch(
    *,
    source: LaunchSource,
    search: SavedSearch | None = None,
    mode: LaunchMode | None = None,
    rejection_reply_text: str | None = None,
    max_scan_depth: int | None = None,
) -> TaskLaunch:
    """Launch a 仅沟通 rejection cleanup.

    Rejection cleanup is keyword-independent, so it carries resolved triage settings
    rather than a search strategy. ``mode=None`` lets the configured ``chat.dry_run``
    win — the convention the searches-page trigger already used and the reason the
    other two callers should stop stating a ``dry_run`` of their own.
    """
    ack = resolve_chat_acknowledgment_settings()

    payload: dict[str, Any] = {
        "dry_run": ack.dry_run if mode is None else mode == LaunchMode.DRAFT,
        "rejection_reply_text": (
            ack.rejection_reply_text if rejection_reply_text is None else rejection_reply_text
        ),
        "max_scan_depth": ack.max_scan_depth if max_scan_depth is None else int(max_scan_depth),
    }
    if search is not None:
        payload["saved_search_id"] = search.id
        payload["search_id"] = search.id
        payload["search_name"] = search.name
    return TaskLaunch(task_type=TaskType.CHECK_CHAT, payload=payload, source=source)


def build_login_diagnostic_launch(*, source: LaunchSource) -> TaskLaunch:
    """Launch the CHECK_LOGIN probe.

    It carries provenance and nothing else: the handler reads no payload at all, and the
    ``mode: "diagnostic"`` key it used to be given had no reader — the diagnostic intent
    is provenance, which the record now carries as an attribute.
    """
    return TaskLaunch(task_type=TaskType.CHECK_LOGIN, payload={}, source=source)


def build_launch(
    kind: TaskKind,
    *,
    source: LaunchSource,
    search: SavedSearch | None = None,
    mode: LaunchMode | None = None,
    **kwargs: Any,
) -> TaskLaunch:
    """The one entry point: turn a launch request into a validated task."""
    if kind == TaskKind.SEARCH:
        if search is None:
            raise LaunchContractError("a search launch requires a SavedSearch")
        return build_search_launch(search, source=source, mode=mode, **kwargs)
    if kind == TaskKind.CHAT_CLEANUP:
        return build_chat_cleanup_launch(source=source, search=search, mode=mode, **kwargs)
    if kind == TaskKind.LOGIN_DIAGNOSTIC:
        return build_login_diagnostic_launch(source=source)
    raise LaunchContractError(f"unknown task kind {kind!r}")


def coerce_source(raw: Any) -> LaunchSource:
    """A task record's provenance, defaulting to ``manual`` for legacy tasks.

    Tasks created before provenance existed carry no ``source``; they are treated as
    manual so the startup sweep never reclaims a task whose origin it cannot prove.
    """
    try:
        return LaunchSource(str(raw))
    except ValueError:
        return LaunchSource.MANUAL
